RFC-0040-nested-device-mesh.md

```diff
@@ -0,0 +1,123 @@
+# (Dynamic)NestedDeviceMesh
+
+**Authors:**
+* @Jackmin801
+* @samsja 
+* @JohannesHa
+
+## **Summary**
+The current abstraction for managing process groups in torch distributed is DeviceMesh, which provides a useful model for defining N-dimensional parallelisms.
+However, the current implementation has some limitations that make it unsuitable for dynamic heterogenous training workloads.
+The first limitation is that the DeviceMesh needs to be cuboid, assuming homogeneity among processes.
+This restriction limits its applicability for heterogeneous training workloads, where different devices might have varying workloads that do not map neatly onto a cuboid mesh.
+The second limitation is that the world size needs to be known at start time. This results in an inability to create dynamic training runs where groups of processes join and leave the training without restarting all the processes in the training run.
+
+To address the first limitation, we propose the introduction of a NestedDeviceMesh abstraction.
+This new abstraction allows for more flexibility by allowing nesting of DeviceMesh instances, an element of in a DeviceMesh can be a DeviceMesh, each potentially with different shapes and sizes.
+This would allow users to model more complex parallelism strategies, such as those required by heterogeneous workloads.
+
+To address the second limitation, we propose the process group creation API be rewritten to support dynamic world size by periodically polling the c10d store for the current world size and recreating the nccl communicators in the case of a change.
+
+## **Motivation**
+This proposal is motivated by the need to support dynamic mixed hardware in OpenDiLoCo.
```

Comment:
we should frame it less towards open diloco and more towards decentralized training. As this would be useful as well for swarm and other

Reply:
yea agree. throughout the doc, OpenDiLoCo is used to mean hierarchical sgd in heterogenous setting for lack of better terminology / wording

Reply:
Yess, haven't read too many other PyTorch RFCs as you guys but would assume we need a bit more general motivation. Replace OpenDiLoCo with "distributed training in heterogeneous setting" and add a bunch of citations of work in that direction (Maybe even include the multi-datacenter training post from semianalysis)